std/experimental/normint.d

```diff
@@ -0,0 +1,571 @@
+
+        static assert(NormalizedInt!ubyte(0x40) % 2 == 0);
+        static assert(NormalizedInt!ubyte(0xFF) % 2 == 1);
+    }
+
+    NormalizedInt!I opBinary(string op, F)(F rh) const if(isFloatingPoint!F && (op == "*" || op == "/" || op == "%"))
+    {
+        return NormalizedInt!I(mixin("cast(F)this " ~ op ~ " rh"));
+    }
+    unittest
+    {
+        static assert(NormalizedInt!ubyte(0x40) * 2.0 == 0x80);
+        static assert(NormalizedInt!ubyte(0x80) * 2.0 == 0xFF);
+        static assert(NormalizedInt!ubyte(0xFF) * 2.0 == 0xFF);
+        static assert(NormalizedInt!byte(32) * 2.0 == 64);
+        static assert(NormalizedInt!byte(64) * 2.0 == 127);
+        static assert(NormalizedInt!byte(127) * 2.0 == 127);
+        static assert(NormalizedInt!byte(-32) * 2.0 == -64);
+        static assert(NormalizedInt!byte(-64) * 2.0 == -127);
+        static assert(NormalizedInt!byte(-127) * 2.0 == -127);
+        static assert(NormalizedInt!byte(-32) * -2.0 == 64);
+        static assert(NormalizedInt!byte(-64) * -2.0 == 127);
+
+        static assert(NormalizedInt!ubyte(0x40) * 0.5 == 0x20);
+        static assert(NormalizedInt!ubyte(0x80) * 0.5 == 0x40);
+        static assert(NormalizedInt!ubyte(0xFF) * 0.5 == 0x80);
+        static assert(NormalizedInt!byte(32) * 0.5 == 16);
+        static assert(NormalizedInt!byte(64) * 0.5 == 32);
+        static assert(NormalizedInt!byte(127) * 0.5 == 64);
+        static assert(NormalizedInt!byte(-32) * 0.5 == -16);
+        static assert(NormalizedInt!byte(-64) * 0.5 == -32);
+        static assert(NormalizedInt!byte(-127) * 0.5 == -64);
+        static assert(NormalizedInt!byte(-32) * -0.5 == 16);
+        static assert(NormalizedInt!byte(-64) * -0.5 == 32);
+
+        static assert(NormalizedInt!ubyte(0xFF) / 2.0 == 0x80);
+        static assert(NormalizedInt!ubyte(0xFF) % 0.5 == 0);
+    }
+
+    NormalizedInt!I opBinary(string op)(NormalizedInt!I rh) const if(op == "|" || op == "&" || op == "^")
+    {
+        return NormalizedInt!I(cast(I)(mixin("value " ~ op ~ " rh.value")));
+    }
+    unittest
+    {
+        static assert((NormalizedInt!uint(0x80) | NormalizedInt!uint(0x08)) == 0x88);
+        static assert((NormalizedInt!uint(0xF0) & NormalizedInt!uint(0x81)) == 0x80);
+        static assert((NormalizedInt!uint(0x81) ^ NormalizedInt!uint(0x80)) == 0x01);
+    }
+
+    NormalizedInt!I opBinary(string op)(int rh) const if(op == "<<" || op == ">>" || op == ">>>")
+    {
+        return NormalizedInt!I(cast(I)(mixin("value " ~ op ~ " rh")));
+    }
+    unittest
+    {
+        static assert(NormalizedInt!uint(0x08000000) << 2 == 0x20000000);
+        static assert(NormalizedInt!int(0x80000000) >> 7 == 0xFF000000);
+        static assert(NormalizedInt!int(0x80000000) >>> 7 == 0x01000000);
+    }
+
+    bool opEquals(NormalizedInt!I rh) const
+    {
+        return value == rh.value;
+    }
+    bool opEquals(T)(T rh) const if(isIntegral!T)
+    {
+        return value == rh;
```

Comment:
Does it make sense to clamp `rh` to the NormalizedInt range? Or maybe limit opEquals to the `I` type? I think it should at least be documented though that comparing with a value outside of the range will return false.